processing/src/main/java/org/apache/druid/segment/IndexIO.java

```diff
@@ -682,6 +689,102 @@ public Metadata getMetadata()
       return index;
     }
 
+    private Map<String, Supplier<ColumnHolder>> readProjectionColumns(
+        ObjectMapper mapper,
+        SegmentLazyLoadFailCallback loadFailed,
+        AggregateProjectionSpec projectionSpec,
+        SmooshedFileMapper smooshedFiles,
+        Map<String, Supplier<ColumnHolder>> columns,
+        Interval dataInterval
+    ) throws IOException
+    {
+      final Map<String, Supplier<ColumnHolder>> projectionColumns = new LinkedHashMap<>();
+      for (DimensionSchema groupingColumn : projectionSpec.getGroupingColumns()) {
+        final String smooshName = Projections.getProjectionSmooshV9FileName(projectionSpec, groupingColumn.getName());
+        final ByteBuffer colBuffer = smooshedFiles.mapFile(smooshName);
+
+        if (columns.containsKey(groupingColumn.getName())) {
+          final String internedColumnName = SmooshedFileMapper.STRING_INTERNER.intern(groupingColumn.getName());
+          projectionColumns.put(internedColumnName, Suppliers.memoize(
```

Comment:
This seems to mostly be a copy of code in `registerColumnHolder` -- can it be refactored to avoid the copy?

It seems like the difference is `serde.read` vs `serde.readProjection`. Do we need `readProjection` really? I mean does it work just as well to add a `@Nullable` parent ref to `serde.read`?

Reply:
reworked to share method after consolidated read methods